src/common/java/net/minecraftforge/gradle/common/legacy/LegacyExtension.java

```diff
@@ -0,0 +1,141 @@
+        final boolean shouldFixClasspath = config.getFixClasspath().get();
+        final boolean shouldExtractMappings = config.getExtractMappings().get();
+
+        if(shouldFixClasspath) {
+            LOGGER.info("LegacyExtension: Fixing classpath");
+            // create a singleton collection from the jar task's output
+            final FileCollection jar = project.files(project.getTasks().named("jar"));
+
+            minecraft.getRuns().stream()
+                    // get all RunConfig SourceSets
+                    .flatMap(runConfig -> runConfig.getAllSources().stream())
+                    .distinct()
+                    // replace output directories with the jar artifact on each SourceSet's classpath
+                    .forEach(sourceSet -> sourceSet.setRuntimeClasspath(sourceSet.getRuntimeClasspath().minus(sourceSet.getOutput()).plus(jar)));
+        }
+
+        if (shouldExtractMappings) {
+            LOGGER.info("LegacyExtension: Extracting Mappings");
+            // Extracts CSV mapping files to a folder in the build directory for further use
+            // by the <code>net.minecraftforge.gradle.GradleStart.csvDir</code> property.
+            final ExtractZip extractMappingsTask = project.getTasks().create("extractMappings", ExtractZip.class, t -> {
+                t.getZip().fileProvider(project.provider(() -> {
+                    // create maven dependency for mappings
+                    String MAPPING_DEP = "net.minecraft:mappings_{CHANNEL}:{VERSION}@zip";
+                    String coordinates = MAPPING_DEP.replace("{CHANNEL}", minecraft.getMappingChannel().get()).replace("{VERSION}", minecraft.getMappingVersion().get());
+
+                    Dependency dep = project.getDependencies().create(coordinates);
+                    // download and cache the mappings
+                    return project.getConfigurations().detachedConfiguration(dep).getSingleFile();
+                }));
+                // mappings are extracted to <root>/build/<taskName> by default
+                t.getOutput().convention(project.getLayout().getBuildDirectory().dir(t.getName()));
+            });
+            final File csvDir = extractMappingsTask.getOutput().get().getAsFile();
+
+            // attach the csvDir location property to each run configuration
+            minecraft.getRuns().configureEach(run -> run.property("net.minecraftforge.gradle.GradleStart.csvDir", csvDir));
+
+            // execute extractMappings before each run task
+            project.getTasks().getByName("prepareRuns").dependsOn(extractMappingsTask);
+        }
+    }
+
+    public LegacyExtension(Project project) {
+        Provider<Boolean> isLegacy = project.provider(() -> {
+            final MinecraftVersion version = MinecraftVersion.from((String) (project.getParent() != null ? project.getParent() : project).getExtensions().getExtraProperties().get("MC_VERSION"));
```

Comment:
This is really annoying for projects with subprojects. Please check if the current project has the property and if it doesn't only then query it from the root project.